src/nubank/auth/mod.rs

```diff
@@ -0,0 +1,182 @@
+}
+
+#[derive(Debug, Clone)]
+pub struct AuthData {
+    access_token: String,
+    feed_url: String,
+    bills_url: String,
+    customer_url: String,
+    query_url: String,
+    revoke_token_url: String,
+}
+
+pub async fn authenticate(
+    path: String,
+    cpf: String,
+    password: String,
+) -> Result<AuthData, AuthError> {
+    let url = get_url("token".to_string()).await?;
+
+    let id = get_identity(path)?;
+    let client = build_client(id)?;
+
+    let payload = build_payload(cpf, password);
+
+    let response = make_auth_request(client, url, payload).await?;
+    let auth_data_dto = read_request_output(response)?;
+    let auth_data = build_auth_data_obj(auth_data_dto)?;
+
+    Ok(auth_data)
+}
+
+async fn get_url(name: String) -> Result<String, AuthError> {
+    let discovery_app_url = "https://prod-s0-webapp-proxy.nubank.com.br/api/app/discovery";
+
+    let urls_str = reqwest::get(discovery_app_url)
+        .await
+        .map_err(|x| AuthError::DiscoveryRequestFailed(x))?
+        .text()
+        .await
+        .map_err(|x| AuthError::DiscoveryRequestDecodingFailed(x))?;
+
+    let urls = serde_json::from_str::<Value>(&*urls_str)
+        .map_err(|x| AuthError::DiscoveryJsonConversionFailed(x))?;
+
+    match urls.get(&name) {
+        Some(v) => Ok(v.to_string().replace("\"", "")),
+        _ => return Err(AuthError::UrlNameNotFoundInDiscoveryJson(name)),
+    }
+}
+
+fn get_identity(path: String) -> Result<reqwest::Identity, AuthError> {
+    let mut buf = Vec::new();
+    let _ = File::open(path)
+        .map_err(|x| AuthError::CannotReadCertificate(x))?
+        .read_to_end(&mut buf);
+    let id = reqwest::Identity::from_pkcs12_der(&buf, "")
+        .map_err(|x| AuthError::CannotInterpretCertificate(x))?;
+    Ok(id)
+}
+
+fn build_client(id: reqwest::Identity) -> Result<reqwest::Client, AuthError> {
+    reqwest::Client::builder()
+        .identity(id)
+        .build()
+        .map_err(|x| AuthError::CannotBuildClient(x))
+}
+
+fn build_payload(cpf: String, password: String) -> String {
+    json!(
+        {
+            "grant_type": "password",
+            "client_id": "legacy_client_id",
+            "client_secret": "legacy_client_secret",
+            "login": cpf,
+            "password": password
+        }
+    )
+    .to_string()
+}
+
+async fn make_auth_request(
+    client: reqwest::Client,
+    url: String,
+    payload: String,
+) -> Result<String, AuthError> {
+    let result = client
+        .post(url)
+        .apply_default_header()
+        .body(payload)
+        .send()
+        .await
+        .map_err(|x| AuthError::AuthRequestFailed(x))?
+        .text()
+        .await
+        .map_err(|x| AuthError::AuthRequestDecodingFailed(x))?;
+
+    Ok(result)
+}
+fn read_request_output(result: String) -> Result<AuthDataDTO, AuthError> {
+    let obj = serde_json::from_str::<AuthDataDTO>(&*result)
+        .map_err(|x| AuthError::AuthJsonConversionFailed(x))?;
```

Comment:
You don't need to use error propagation here, just return the result.

Reply:
Done.